### DISCUSSION TITLE
Resend confirmation email

### DISCUSSION THREAD

**Discussion by b2m9** [supabase/supabase#3526] [Questions] [answered]
Is there an option to resend the confirmation email when a user has signed up for a new account?

I don't find a good reference for the admin API, but my best guess was using `auth.api.generateLink` like this:

```js
const { error, data } = await supabaseAdmin.auth.api.generateLink(
    'signup', email, { redirecTo: 'url' });
```

It doesn't return any `error` or `data` nor does it trigger an email.

Any ideas are highly appreciated. Thanks.

**Comment by silentworks** [comment]
That would not send an email, it would generate a link which you could then send by your own means. I also think this function needs to be used with the `service_role` `secret`.

Note: if you are using the `service_role` `secret` make sure it's on the server-side and not on your client-side as this gives a user access to everything.

**Reply by b2m9** [reply]
Got it, `generateLink` only creates the link for me.

That being said, is there a method that triggers the confirmation email? You already have methods for the other use cases (`sendMagicLinkEmail`, `resetPasswordForEmail`, `inviteUserByEmail`). However, I couldn't find anything that fits the use case.

**Reply by silentworks** [reply]
At the moment I don't think there is, but you can open an issue on GitHub asking for this feature to be added.

**Comment by b2m9** [comment]
As of October 2021, it is not possible to resend the confirmation email. An [issue has been opened in July 2021](https://github.com/supabase/supabase/issues/2210) to address this.

The only way is to create the a link with `auth.api.generateLink` and send the email yourself (e.g. via Postmark, Sendgrid, etc.).

Thanks @silentworks for the assistance. :+1:

**Reply by bdthombre** [reply]
Is there any way to resent the confirmation email.

**Reply by b2m9** [reply]
Until aforementioned issue https://github.com/supabase/gotrue/issues/312 has been resolved, no. You can still use `auth.api.generateLink` and send an email yourself.

**Reply by arpitdalal** [reply]
How do I use `auth.api.generateLink`? It gives me `Session | User | null`, how do I extract the confirmation link from that?

**Reply by arpitdalal** [reply]
nvm, figured it out. The returned object has `action_link` on it but the ts types are way off.

**Reply by arpitdalal** [reply]
@b2m9 when passing an email address that has signed up just now to `auth.api.generateLink("signup", email)` gives me an error saying "A user with this email address has already been registered" Of course, there is already a user with that email, but the email is not verified so I want to resend the verification email. Does it not work that way?

**Comment by maxibenner** [comment]
@b2m9 I think I found a workaround. Please let me know of any side effects I might have missed.

The initial confirmation email is sent on signup and there is no way in the client side sdk to resend it. However, if you keep track of the email address and password (in a variable; I'm not advocating writing the plain password anywhere else), you can just sign up again using the same credentials. As long as the email has not been confirmed, yet, it just overwrites the initial user entry and "resends" the confirmation email. The only limitation is a mandatory one minute timeout before signing up with the same email address again.

**Comment by kangmingtay** [comment]
Just to clarify, are you looking to resend the confirmation email as an admin user to your users who have signed up but are not yet confirmed? 

Else, the user can just sign up again which results in the confirmation email being resent.

**Reply by silentworks** [reply]
Yes as an admin user to users who have signed up but have not confirmed as yet. It would be good to have a function for this in the JS library instead of the current method of generating the link and then sending with my own email mechanism.

**Reply by kangmingtay** [reply]
Hmm isn't this just using the invite endpoint though? IIRC you can send multiple invites to a user as an admin

**Reply by chipilov** [reply]
@kangmingtay Not sure if @silentworks's scenario is the same, but due to the fact that e-mail delivery is NOT always reliable (e.g. you might hit some kind of rate limiting during a burst), it is common for sign-up pages to have a button similar to "Resend link" which the user can press if the confirmation e-mail does NOT arrived within reasonable time (e.g. 1 minute).

Implementing this button in Supabase seems surprisingly complex for the following reasons:
  - As this discussion mentions, currently this ONLY possible by writing server code that executes 2 steps: (1) generate the confirmation link and (2) use an external email provider to send that link to the user;
  - To make matters worse, it's hard to make the email that is initially sent by Supabase on sign-up look exactly the same as the email sent when the user presses the "Resend link" button. I've explained this in more detail here: https://github.com/supabase/gotrue/issues/304#issuecomment-1012302982

I see 3 approaches that could alleviate this issue (there might be another approach that I cannot think of):
  - Allow enabling e-mail confirmation WITHOUT actually sending e-mails. That is, the logic in the GoTrue server would ONLY generate the confirmation token without actually sending an e-mail and a developer-defined trigger would actually send the confirmation. There is a tricky problem to solved with this approach, though - calling the generate_link API from the trigger currently does NOT work, because the trigger runs BEFORE the transaction is finished and so the generate_link API cannot find the new user and returns HTTP 422;
  - Allow invoking HTTP endpoints on signup/password reset, etc. That way, users who use external e-mail providers (which is recommended in the Supabase docs) can just provide the HTTP endpoints for send e-mails supplied by their e-mail providers. Of course, these links need to be parameterized;
  - as suggested by @silentworks, introduce a method which re-sends the confirmation email. This would probably most of the pain, but the problem with this approach is that it still does NOT solve the problem of having to maintain 2 sets of templates (one in Supabase and one in the e-mail provider of choice);

**Comment by kbsali** [comment]
I have another use case related to all this : I would like to signup users WITHOUT having to confirm their emails in the first place.
To do so you can easily disable **email confirmations** in your project settings.
Nevertheless I would still like to validate the user's email by sending him a confirmation email "programatically" calling  `generateLink('signup'...`, but it won't work. When disabling **email confirmations**, the accounts created are automatically confirmed.

Any idea how to work around this issue?
Thanks

**Reply by KhaledGabr** [reply]
exactly! that's what I am wondering as well. Seems like we need 2 checkboxes.
[Require email confirmations]
[Send email Confirmations Emails]

If  **require email confirmation** is set to true, it would only require it without automatically sending emails. 
It will expect users to programmatically call the generate link and send the email to users. 

But if both are set to true, then it will require and send confirmation email to users.

**Reply by ZhenFTW** [reply]
This is what I was looking for as well. I agree with this. I think having a new type 'signin' or 'resend' in generateLink is essential to send a new verification link for already created user users. The ability to verify a user only after signup is what supabase auth is missing right now. Gotrue is pretty much amazing and can almost do everything out there, but this is one of the limitation that I find that the library is missing right now. I hope the team find a way to solve this.

**Comment by madebyfabian** [comment]
For everyone stumbling on this. I solved this (whithout using custom emails) by moving the signup logic completely on the server by using `supabaseAdmin.auth.admin.createUser`. 

And regardless if this request was successful (user was not registered, now is registered) or if the request fails (because user already is registered), I always do a `supabaseAdmin.auth.admin.inviteUserByEmail`. Since this is currently the only method which actually sends an email, without having to use custom emails and `generateLink`. 
The `verifyUser` call then sends a "you've beein invited" email. Still not perfect, but it works and it's still safe since all happens on server-side. Since the client never gets to know if there actually already was a user or not.

**Be aware that this seems to only work on accounts that have not password set (oAuth/Magic Link).**

**Comment by florianwalther-private** [comment]
I agree that this should be fixed. What I am doing to work around this problem is using `signInWithOtp` instead of normal password sign up, because you can send as many of these as you want. I then let the user set the password after they entered the OTP.

**Comment by tholder** [comment]
This was pretty confusing for us, but our findings show that if you call signUpWithOtp https://supabase.com/docs/reference/javascript/auth-signinwithotp this does the email confirmation if they are not confirmed.

**Comment by abrakazinga** [comment]
the required gotrue feature pr was merged last week into gotrue-js. see: https://github.com/supabase/gotrue-js/pull/631

**Comment by jmshgs** [comment]
you can now use 
```
const { data, error } = await supabase.auth.resend({
  type: 'signup',
  email: 'email@example.com'
})
```

it looks like to send a signup email. docs: https://supabase.com/docs/reference/javascript/auth-resend

**Comment by kangmingtay** [comment, marked as answer]
hi everyone, this is now possible by doing `supabase.auth.resend(...)` - check it out here: https://supabase.com/docs/reference/javascript/auth-resend

**Reply by landmann** [reply]
I'm sorry but it still does not send the e-mail.

I'm trying this
```
 const { data, error } = await supabase.auth.resend({
     type: "email_change",
     email: email,
});
```
but the inbucket is not receiving any emails.  I can see `auth.identities` recorded the email change the first time around....

<img width="968" alt="image" src="https://github.com/supabase/supabase/assets/12756081/32df756d-7f2d-466b-8004-030e5eef7c26">

**Reply by Kylar13** [reply]
The provided method doesn't seem to send any emails, but requests complete as if successful:

![image](https://github.com/supabase/supabase/assets/7138244/91a80791-0458-445b-aa35-ee0d6a21682f)

Even after the screenshot seen in logs, we don't get any new emails nor our resend.com dashboard shows any new requests

**Reply by kangmingtay** [reply]
@Kylar13 @landmann the method only resends the email if there was an email of the specific type requested before

**Reply by kangmingtay** [reply]
@landmann in your case, the first call to send the magic link failed with a `User already registered` error which is why the resend action doesn't send out anything

**Reply by Kylar13** [reply]
> @Kylar13 @landmann the method only resends the email if there was an email of the specific type requested before

That's what I'm doing though.

I've implemented signInWithOTP() and in the "Verify Code" screen I've added a "Resend code" button that calls resend().

Am I missing something?

**Reply by kangmingtay** [reply]
@Kylar13 check if there's an error returned by your `signInWithOtp()` method by looking at your network tab and observing the response returned by `/auth/v1/otp`

**Reply by Kylar13** [reply]
> @Kylar13 check if there's an error returned by your `signInWithOtp()` method by looking at your network tab and observing the response returned by `/auth/v1/otp`

Nono, signInWithOtp works fine and I get the code, but I was trying to test the "Resend code" functionality and I would not receive any new email nor see anything in the resend.com dashboard

**Reply by kangmingtay** [reply]
@Kylar13 to resend OTPs originating from the `signInWithOtp()` call, you can just call `signInWithOtp()` again - as a reference, we also mention this in our docs that the resend method is only applicable to signup, email change or phone change (https://supabase.com/docs/reference/javascript/auth-resend)

**Reply by Tultra** [reply]
When using supabase.auth.resend(type: "signup", email:"user@email"}) to resend an e-mail confirmation, I receive as response a json object full of nulls that end up overriding the session stored locally (setting it to null). The json received as response is the following: {"data": {"session": null, "user": null}, "error": null}. Am I missing something?

**Reply by neelbrahmakshatriya** [reply]
Hi @Tultra, did you find any solution to this? I'm also getting the same JSON response with all the values as null.

**Reply by thomas-j-back** [reply]
I am getting this response too using resend, I also have been unable to verify it actually sends the email.

**Comment by dejongyeong** [comment]
Hi everyone, just wondering if would it be possible to generate a resend confirmation email link without the password? something like below:

```
const {data, error} = await supabase.auth.admin.generateLink({
   type: 'signup',
   email: 'email@example.com',
   options: {redirectTo: `localhost:3000`}
});
```

thanks!

**Reply by kangmingtay** [reply]
@dejongyeong have you seen https://supabase.com/docs/reference/javascript/auth-admin-generatelink? i think what you're looking for is `type: magiclink` instead of signup if you want the user to use a passwordless login method

**Reply by dejongyeong** [reply]
@kangmingtay, by any chance to generate a resend confirmation email link? At first, the admin will generate a sign-up link and send it to the user email via a third-party email provider, then if we want to resend the confirmation email link via a third-party email provider as well, which option should we choose so that we can omit out the password when generate a resend confirmation link?

Not sure if ```type: magiclink``` suits.

**Reply by kangmingtay** [reply]
@dejongyeong  why not just generate another signup link and send it?

**Reply by dejongyeong** [reply]
@kangmingtay yes, in the end, that is what I tried and uses the ```type: 'invite'``` to resend an invite.

**Reply by kangmingtay** [reply]
yes, that makes sense, i don't think there's an actual use case to support generating a resend confirmation email link here

**Comment by neelbrahmakshatriya** [comment]
I have written the following code in my server actions file for resending the confirmation email:

```tsx
'use server'

import { createServiceRoleClient } from '@/utils/supabase/server'

export const verifyEmail = async (email: string) => {
  const supabase = createServiceRoleClient()

  const response = await supabase.auth.resend({
    type: 'signup',
    email,
    options: {
      emailRedirectTo: `${process.env.NEXT_PUBLIC_APP_URL}/auth/callback?next=/dashboard`
    }
  })

  console.log(response)

  return response
}
```

When I log the response, I keep seeing this in my terminal:

```
{ data: { user: null, session: null }, error: null }
```

I'm stuck and can't figure out where I'm going wrong. Any guidance would be amazing! 🙏

Thanks in advance!

**Reply by bdhamithkumara** [reply]
const { data, error } = await (await supabase).auth.resend({
        type: 'signup',
        email: formData?.email,
        options: {
          emailRedirectTo: `${url}join?is_verified=true&existing=${found}`
        }
      })
i used this method it work for me

**Reply by wu0407** [reply]
I see the code all return null except error
  
 https://github.com/supabase/auth-js/blob/dcc2b31fc78be803be09110842ce4312d6b1a2ba/src/GoTrueClient.ts#L1109-L1145

**Reply by adds08** [reply]
Same error
